### PR TITLE
feat(plasma-hope,plasma-web,plama-b2c): Add `TextArea` union

### PR DESCRIPTION
**Pull request by neretin-trike**
:baby_chick: Download canary assets:

[colors--canary.405.4439006335.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.405.4439006335.xml)  
[color_metro_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_ios-swift--canary.405.4439006335.swift)  
[color_metro_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_kotlin--canary.405.4439006335.kt)  
[color_metro_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_metro_react-native--canary.405.4439006335.ts)  
[color_sberHealth_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_ios-swift--canary.405.4439006335.swift)  
[color_sberHealth_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_kotlin--canary.405.4439006335.kt)  
[color_sberHealth_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_react-native--canary.405.4439006335.ts)  
[color_sbermarket_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.405.4439006335.swift)  
[color_sbermarket_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.405.4439006335.kt)  
[color_sbermarket_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.405.4439006335.ts)  
[color_sberprime_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.405.4439006335.swift)  
[color_sberprime_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.405.4439006335.kt)  
[color_sberprime_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.405.4439006335.ts)  
[color_selgros_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_ios-swift--canary.405.4439006335.swift)  
[color_selgros_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_kotlin--canary.405.4439006335.kt)  
[color_selgros_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_selgros_react-native--canary.405.4439006335.ts)  
[color_smbusiness_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_ios-swift--canary.405.4439006335.swift)  
[color_smbusiness_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_kotlin--canary.405.4439006335.kt)  
[color_smbusiness_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_smbusiness_react-native--canary.405.4439006335.ts)  
[PlasmaTokensColor--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.405.4439006335.swift)  
[shadow_sbermarket_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.405.4439006335.ts)  
[typo_mage_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.405.4439006335.swift)  
[typo_mage_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.405.4439006335.kt)  
[typo_mage_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.405.4439006335.ts)  
[typo_plasma_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.405.4439006335.swift)  
[typo_plasma_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.405.4439006335.kt)  
[typo_plasma_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.405.4439006335.ts)  
[typo_ruler_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.405.4439006335.swift)  
[typo_ruler_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.405.4439006335.kt)  
[typo_ruler_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.405.4439006335.ts)  
[typo_sage_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.405.4439006335.swift)  
[typo_sage_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.405.4439006335.kt)  
[typo_sage_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.405.4439006335.ts)  
[typo_sbermarket_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.405.4439006335.swift)  
[typo_sbermarket_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.405.4439006335.kt)  
[typo_sbermarket_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.405.4439006335.ts)  
[typo_soulmate_ios-swift--canary.405.4439006335.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.405.4439006335.swift)  
[typo_soulmate_kotlin--canary.405.4439006335.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.405.4439006335.kt)  
[typo_soulmate_react-native--canary.405.4439006335.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.405.4439006335.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.163.0-canary.405.4439006335.0
  npm install @salutejs/plasma-colors@0.6.0-canary.405.4439006335.0
  npm install @salutejs/plasma-core@1.105.0-canary.405.4439006335.0
  npm install @salutejs/plasma-hope@0.21.0-canary.405.4439006335.0
  npm install @salutejs/plasma-icons@1.133.0-canary.405.4439006335.0
  npm install @salutejs/plasma-temple@1.144.0-canary.405.4439006335.0
  npm install @salutejs/plasma-tokens-b2b@1.13.0-canary.405.4439006335.0
  npm install @salutejs/plasma-tokens-b2c@0.22.0-canary.405.4439006335.0
  npm install @salutejs/plasma-tokens-web@1.28.0-canary.405.4439006335.0
  npm install @salutejs/plasma-tokens@1.42.0-canary.405.4439006335.0
  npm install @salutejs/plasma-ui@1.177.0-canary.405.4439006335.0
  npm install @salutejs/plasma-web@1.188.0-canary.405.4439006335.0
  npm install @salutejs/plasma-cy-utils@0.49.0-canary.405.4439006335.0
  npm install @salutejs/plasma-sb-utils@0.103.0-canary.405.4439006335.0
  npm install @salutejs/plasma-tokens-android@2.49.0-canary.405.4439006335.0
  npm install @salutejs/plasma-tokens-ios-swift@2.49.0-canary.405.4439006335.0
  # or 
  yarn add @salutejs/plasma-b2c@1.163.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-colors@0.6.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-core@1.105.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-hope@0.21.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-icons@1.133.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-temple@1.144.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-tokens-b2b@1.13.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-tokens-b2c@0.22.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-tokens-web@1.28.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-tokens@1.42.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-ui@1.177.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-web@1.188.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-cy-utils@0.49.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-sb-utils@0.103.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-tokens-android@2.49.0-canary.405.4439006335.0
  yarn add @salutejs/plasma-tokens-ios-swift@2.49.0-canary.405.4439006335.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-405/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out comparison report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-405/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-405/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-405/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-405/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-405/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-405/temple-storybook/

File: packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    position: absolute;
+
+    width: ${({ width }) => width}px;
+`;
+
+/**
+ * Поле ввода многострочного текста.
+ */
+export const TextAreaB2C = ({
+    // label,
```
Comment (Yakutoc) [inline packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx:109]: Почему закомментировано?

**Comment by Yakutoc** [comment]
@neretin-trike Я же правильно понял, что решили баги которые ты нашел по этому компоненту вынести в отдельные PR?

File: packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx (hunk 1 above)
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx:109]: удолю

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-405/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-405/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-405/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-405/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-405/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-405/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-405/temple-storybook/

File: packages/plasma-core/src/components/TextArea/TextArea.component-test.tsx [hunk 2]
```diff
@@ -52,8 +64,8 @@ describe('plasma-core: TextArea', () => {
                 <TextArea value="Value" placeholder="Placeholder" helperText="Helper text" status="success" />
                 <SpaceMe />
                 <TextArea value="Value" placeholder="Placeholder" helperText="Helper text" status="warning" />
-                <SpaceMe />
-                <TextArea value="Value" placeholder="Placeholder" helperText="Helper text" status="error" />
+                {/* <SpaceMe />
```
[the remark this answers is not in the document]
Comment (neretin-trike) [reply packages/plasma-core/src/components/TextArea/TextArea.component-test.tsx:67]: это забыл раскоментировать, когда отлаживался

File: packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                {hasHelper && (
+                    <StyledFieldHelpers width={width} id={id ? `${id}-helper` : undefined}>
+                        {leftHelper && <TextFieldHelper>{leftHelper}</TextFieldHelper>}
+                        {rightHelper && <TextFieldHelper>{rightHelper}</TextFieldHelper>}
+                    </StyledFieldHelpers>
+                )}
+            </FieldWrapper>
+        </TextAreaWrapper>
+    );
+};
```
[the remark this answers is not in the document]
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx:165]: С одной стороны да, с другой не совсем. Я, помимо самих представлений, вынес логику компонента в один файл и таким образом этот подход улучшит поддержку самого компонента:
1. Если сломается логика - её нужно будет изменить в одном файле,
2. Если изменится дизайн у какой-то одной темы (например b2c), это можно будет легко изменить (за счёт читаемости и прозрачности реализации) в файле, который лежит рядом с другими представлениями

Но так же это накладывает некоторые ограничения: если появятся новая тема, необходимо будет добавить новое представление. Такую проблему нужно будет учесть при следующих итерациях работы над компонентами (переезд на новую систему токенов и редизайн)

File: packages/plasma-hope/src/components/TextArea/types.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    /**
+     * Вспомогательный текст снизу слева для поля ввода.
+     */
+    leftHelper?: string;
+    /**
+     * Вспомогательный текст снизу справа для поля ввода.
+     */
+    rightHelper?: string;
+    /**
+     * Возможные размеры компонента (plasma-web: 'l' | 'm').
```
Comment (Yakutoc) [inline packages/plasma-hope/src/components/TextArea/types.ts:21]: @neretin-trike Почему в комментарии указано 2 значения, а их 

`size?: 'l' | 'm' | 's' | 'xs';` ?

File: packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yakutoc) [reply packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx:165]: Я правильно понял что в этом PR решаем задачу через новый APi с prop view(design). 

А уже следующая реализация думаем как сделать чище и проще?

**Review by Yakutoc — APPROVED** [review]

File: packages/plasma-hope/src/components/TextArea/TextArea.tsx [hunk 5]
```diff
@@ -1,120 +1,38 @@
...
-    /**
-     * Вспомогательный текст снизу слева для поля ввода.
-     */
-    leftHelper?: string;
-    /**
-     * Вспомогательный текст снизу справа для поля ввода.
-     */
-    rightHelper?: string;
-}
+import { TextAreaWeb } from './views/web/TextAreaWeb';
```
[the remark this answers is not in the document]
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/TextArea.tsx:5]: Не совсем понял вопроса, а где? 🤔

File: packages/plasma-hope/src/components/TextArea/types.ts (hunk 4 above)
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/types.ts:21]: Там в комментариях написано, што только для plasma-web всего два значения актуальные (хотя на самом деле даже сейчас их изменение ни на что не влияет). Мб не совсем понятно, но это сделано из-за объединения интерфейсов из b2c (там все размеры есть) и web (там только два)

File: packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx:165]: Ну сейчас и так достаточно просто, как раз расписал в каких аспектах. Скорее следующая реализация будет включать попытку объединить в один вью

[the remark this answers is not in the document]
Comment (Yeti-or) [reply packages/plasma-hope/src/components/TextArea/views/b2c/TextAreaB2C.tsx:165]: 👌

File: packages/plasma-hope/src/components/TextArea/types.ts (hunk 4 above)
Comment (Yeti-or) [reply packages/plasma-hope/src/components/TextArea/types.ts:21]: а в итоге-то сколько будет?
мы же объединяем их и API будет одинаковое ? или таки нет?

File: packages/plasma-hope/src/components/TextArea/views/web/TextAreaWeb.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    -webkit-appearance: none; /* для отображение рамок на iOS */
+
+    ${applyInputStyles}
+`;
+
+const StyledFieldHelpers = styled.div<{ width: number }>`
+    display: flex;
+    justify-content: space-between;
+
+    width: ${({ width }) => width}px;
```
Comment (Yeti-or) [inline packages/plasma-hope/src/components/TextArea/views/web/TextAreaWeb.tsx:18]: интересно почему у нас тут пиксели =/

**Review by Yeti-or — APPROVED** [review]
В остальном все ок

File: packages/plasma-hope/src/components/TextArea/views/web/TextAreaWeb.tsx (hunk 6 above)
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/views/web/TextAreaWeb.tsx:18]: честно сказать не помню, возможно мне было сложно в ремах считать когда я динамическую изменяемую ширину делал, а возможно просто пропустил. В целом могу попробовать в ремах тут сделать, единственное, что придётся вычисление делать на каждое изменение ширины

File: packages/plasma-hope/src/components/TextArea/types.ts (hunk 4 above)
Comment (neretin-trike) [reply packages/plasma-hope/src/components/TextArea/types.ts:21]: смотри, в итоге и для plasma-web и plasma-b2c выбрать можно будет все размеры (как раз чтобы один API был). Но для plasma-b2c они будут изменять фактический размер, а для plasma-web нет (так же сейчас и в мастере, и это из-за того, что редизайн не делали). А пометку  `(plasma-web: 'l' | 'm')` я сделал для того, чтобы показать "легаси" апи, может имеет смысл как-то дополнить, чтобы более понятно было

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-405/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-405/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-405/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-405/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-405/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-405/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-405/

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-405/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-405/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-405/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-405/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-405/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-405/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Pull request merged**